src/mqtt.rs

```diff
@@ -42,35 +47,140 @@ impl Mqtt {
+        debug!("Publishing to {topic} with payload {payload}");
+
+        let payload = serde_json::to_string(&payload).unwrap();
+        if let Err(e) = self.client.publish(topic, QoS::AtMostOnce, true, payload) {
+            error!("Failed to publish message: {e}");
+        }
+    }
+
+    fn publish_configs(&mut self, config_topic: &str, sensor_configs: &Vec<SensorConfig>) {
+        // configs let home assistant know what sensors are available and where to find them
+        for sensor_config in sensor_configs {
+            let config_topic = format!("{}/{}/config", config_topic, sensor_config.unique_id);
+            let config_payload = serde_json::to_value(&sensor_config).unwrap();
+            self.publish_json(&config_topic,  config_payload);
+        }
+    }
+
+    fn publish_states(&mut self, hms_state: &HMSStateResponse, state_topic: &str) {
+        // states contain the actual data
+        let json_payload = hms_state.to_json_payload();
+        self.publish_json(state_topic, json_payload);
+    }
 }
 
+
 impl MetricCollector for Mqtt {
     fn publish(&mut self, hms_state: &HMSStateResponse) {
-        debug!("{hms_state}");
-
-        let pv_current_power = hms_state.pv_current_power as f32 / 10.;
-        let pv_daily_yield = hms_state.pv_daily_yield;
-
-        self.client
-            .subscribe("hms800wt2/pv_current_power", QoS::AtMostOnce)
-            .unwrap();
-        match self.client.publish(
-            "hms800wt2/pv_current_power",
-            QoS::AtMostOnce,
-            true,
-            pv_current_power.to_string(),
-        ) {
-            Ok(_) => {}
-            Err(e) => warn!("mqtt error: {e}"),
+        let config_topic = format!("homeassistant/sensor/hms_{}", hms_state.short_dtu_sn());
+        let state_topic = format!("solar/hms_{}/state", hms_state.short_dtu_sn());
+
+        let device_config = hms_state.create_sensor_configs(&state_topic);
+
+        self.publish_configs(&config_topic, &device_config);
+        self.publish_states(hms_state,  &state_topic);
+    }
+}
+
+
+/// `HMSStateResponse` is a struct that contains the data from the inverter.
+///
+/// Provide utility functions to extract data from the struct.
+impl HMSStateResponse {
+    fn get_model(&self) -> String {
+         // TODO: figure out a way to properly identify the model
+        format!("HMS-WiFi")
+    }
+
+    fn get_name(&self) -> String {
+        format!("Hoymiles {} {}", self.get_model(), self.short_dtu_sn())
+    }
+
+    fn short_dtu_sn(&self) -> String {
+        self.dtu_sn[..8].to_string()
+    }
+
+    fn get_total_efficiency(&self) -> f32{
+        let total_module_power: f32 = self.port_state.iter().map(
+            |port| port.pv_power as f32
+        ).sum();
+        if total_module_power > 0.0 {
+            self.pv_current_power as f32 / total_module_power * 100.0
+        } else { 
+            0.0
+        }
+    }
+
+    fn to_json_payload(&self) -> serde_json::Value {
```

Comment:
one thing that I haven't fully understood, yet of this is a json schema that HA requires. Is there any formal requirement, and where can I find some information on this?

Reply:
Home assistant requries two parts: config messages and state messages. The config messages define the sensors and corresponding state topics to listen for state messages. 

The state messages don't require any specific schema. Rather the configuration messages contain a "value_template" (`SensorConfig.value_template`), which is used to extract the state value from a json-encoded message sent to the "state_topic" (`SensorConfig.state_topic`). The template language used is jinja.

We've defined the template for each sensor as `let value_template = format!("{{{{ value_json.{} }}}}", unique_id);`, whereas `value_json` would correspond to the json-serialized `HMSStateResponse` and `unique_id` is defined to match a corresponding attribute in `HMSStateResponse`